dask/dataframe/multi.py

```diff
@@ -919,24 +925,33 @@ def stack_partitions(dfs, divisions, join="outer"):
     dsk = {}
     i = 0
     for df in dfs:
+        df_copy = deepcopy(df)
         # dtypes of all dfs need to be coherent
         # refer to https://github.com/dask/dask/issues/4685
-        if is_series_like(df) and is_series_like(meta):
-            if not df.dtype == meta.dtype and str(df.dtype) != "category":
-                df = df.astype(meta.dtype)
+        if is_dataframe_like(df_copy):
+            shared = df_copy._meta.columns.intersection(meta.columns)
+            for col in shared:
+                if (
+                    df_copy[col].dtype != meta[col].dtype
+                    and str(df_copy[col].dtype) != "category"
+                ):
+                    df_copy[col] = df_copy[col].astype(meta[col].dtype)
```

Comment:
OK, this is the line that's causing issues. Without your `copy` call, we're mutating the caller inplace, which we don't want.

So, what I'd like to see:

1. Just use `df = df.copy()` rather than deepcopy.
2. Only copy when necessary, i.e. only if there are any columns that need to be cast.